### ISSUE TITLE
CAS errors when installing atomic on RedHat EL5 with GCC 4.1.2.

### ISSUE THREAD

**Issue by klynton** [headius/ruby-atomic#24]
Hi,

Trying to install Rails on EL5 isn't working because of the error:

```
Fetching: atomic-1.1.10.gem (100%)
Building native extensions.  This could take a while...
ERROR:  Error installing rails:
    ERROR: Failed to build gem native extension.

    /usr/local/bin/ruby2.0 extconf.rb
creating Makefile

make "DESTDIR="
compiling atomic_reference.c
atomic_reference.c:58:2: error: #error No CAS operation available for this platform
make: *** [atomic_reference.o] Error 1
```

Looking at the macros defined I'm seeing:

```
#define __GNUC__ 4
#define __GNUC_GNU_INLINE__ 1
#define __GNUC_MINOR__ 1
#define __GNUC_PATCHLEVEL__ 2
#define __GNUC_RH_RELEASE__ 54
```

and looking through the docs for that version of GCC I'm seeing the `__sync_bool_compare_and_swap` function that is being called as a supported function here: http://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Atomic-Builtins.html

The same gems install just fine on RedHat EL6 or higher with the following compiler:

```
#define __GNUC__ 4
#define __GNUC_GNU_INLINE__ 1
#define __GNUC_MINOR__ 4
#define __GNUC_PATCHLEVEL__ 7
#define __GNUC_RH_RELEASE__ 3
```

I've tried installing with the latest patch version of Ruby 1.9.3 and Ruby 2.0.0 on both EL5 and EL6.

**Comment by rainhead** [comment]
Embarrassing as it is to admit, we're also on RHEL5 and affected by this issue.

**Comment by rainhead** [comment]
We were able to solve this issue by building with -march=i686.

**Comment by klynton** [comment]
@rainhead was that manually installing the gem after you'd downloaded it or passing something to bundler/gem install?

**Comment by headius** [comment]
I'm willing to incorporate a patch if someone can figure it out, but the C part of this gem is not really my baby.

**Comment by 4min** [comment]
Before the fix gets incorporated, you can use the following workaround:
gem install atomic -- --with-cflags=-march=i686

**Comment by klynton** [comment]
It does indeed work:

```
[testweb@testweb ~]$ gem install atomic -- --with-cflags=-march=i686
Fetching: atomic-1.1.10.gem (100%)
Building native extensions with: '--with-cflags=-march=i686'
This could take a while...
Successfully installed atomic-1.1.10
Parsing documentation for atomic-1.1.10
unable to convert "\xC0" from ASCII-8BIT to UTF-8 for lib/atomic_reference.so, skipping
Installing ri documentation for atomic-1.1.10
1 gem installed
```

**Comment by royisme** [comment]
It works！

**Comment by sanichi** [comment]
I had the same problem on a production machine (Centos 5.9 32 bit) though not on my MacOS dev machine.

sudo gem install atomic -- --with-cflags=-march=i686

did work, but that didn't fix the capistrano deployment, which still tried to re-install the gem, and the fix for that was:

bundle config --global build.atomic --with-cflags=-march=i686

**Comment by klynton** [comment]
@headius: @rainhead has a patch that is passing tests in #26 can it be merged and a new gem released?

**Comment by mamady** [comment]
@4min Thanks for the workaround - lifesaver!

**Comment by headius** [comment]
Fixed by #26.

**Comment by headius** [comment]
atomic-1.1.11 pushed with this fix.
